File: flutter_app/lib/models/reminder.dart
class Reminder {
  final int? id;
  final String reminderType;
  final String scheduledTime;
  final String? content;
  final bool isEnabled;

  Reminder({
    this.id,
    required this.reminderType,
    required this.scheduledTime,
    this.content,
    this.isEnabled = true,
  });

  factory Reminder.fromJson(Map<String, dynamic> json) {
    return Reminder(
      id: json['id'] as int?,
      reminderType: json['reminder_type'] as String,
      scheduledTime: json['scheduled_time'] as String,
      content: json['content'] as String?,
      isEnabled: json['is_enabled'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'reminder_type': reminderType,
      'scheduled_time': scheduledTime,
      'content': content ?? '',
      'is_enabled': isEnabled,
    };
  }

  Reminder copyWith({
    int? id,
    String? reminderType,
    String? scheduledTime,
    String? content,
    bool? isEnabled,
  }) {
    return Reminder(
      id: id ?? this.id,
      reminderType: reminderType ?? this.reminderType,
      scheduledTime: scheduledTime ?? this.scheduledTime,
      content: content ?? this.content,
      isEnabled: isEnabled ?? this.isEnabled,
    );
  }

  bool get isBedtime => reminderType == 'bedtime';
  bool get isPeriodic => reminderType == 'periodic';
}

File: backend/internal/handlers/reminder.go
package handlers

import (
	"net/http"

	"daily-planner-backend/internal/database"
	"daily-planner-backend/internal/models"

	"github.com/gin-gonic/gin"
)

type CreateReminderRequest struct {
	ReminderType  string `json:"reminder_type" binding:"required"`
	ScheduledTime string `json:"scheduled_time" binding:"required"`
	Content       string `json:"content"`
}

type UpdateReminderRequest struct {
	ScheduledTime string `json:"scheduled_time"`
	Content       string `json:"content"`
	IsEnabled     *bool  `json:"is_enabled"`
}

func GetReminders(c *gin.Context) {
	userID := c.GetUint("userID")

	var reminders []models.Reminder
	if err := database.GetDB().Where("user_id = ?", userID).Find(&reminders).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "internal server error"})
		return
	}

	c.JSON(http.StatusOK, reminders)
}

func CreateReminder(c *gin.Context) {
	userID := c.GetUint("userID")

	var req CreateReminderRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "validation error", "details": err.Error()})
		return
	}

	reminder := models.Reminder{
		UserID:        userID,
		ReminderType:  req.ReminderType,
		ScheduledTime: req.ScheduledTime,
		Content:       req.Content,
		IsEnabled:     true,
	}

	if err := database.GetDB().Create(&reminder).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "internal server error"})
		return
	}

	c.JSON(http.StatusCreated, reminder)
}

func UpdateReminder(c *gin.Context) {
	userID := c.GetUint("userID")
	reminderID := c.Param("id")

	var reminder models.Reminder
	if err := database.GetDB().First(&reminder, reminderID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "resource not found"})
		return
	}

	if reminder.UserID != userID {
		c.JSON(http.StatusForbidden, gin.H{"error": "access denied"})
		return
	}

	var req UpdateReminderRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "validation error", "details": err.Error()})
		return
	}

	updates := make(map[string]interface{})
	if req.ScheduledTime != "" {
		updates["scheduled_time"] = req.ScheduledTime
	}
	if req.Content != "" {
		updates["content"] = req.Content
	}
	if req.IsEnabled != nil {
		updates["is_enabled"] = *req.IsEnabled
	}

	if err := database.GetDB().Model(&reminder).Updates(updates).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "internal server error"})
		return
	}

	c.JSON(http.StatusOK, reminder)
}

func DeleteReminder(c *gin.Context) {
	userID := c.GetUint("userID")
	reminderID := c.Param("id")

	var reminder models.Reminder
	if err := database.GetDB().First(&reminder, reminderID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "resource not found"})
		return
	}

	if reminder.UserID != userID {
		c.JSON(http.StatusForbidden, gin.H{"error": "access denied"})
		return
	}

	if err := database.GetDB().Delete(&reminder).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "internal server error"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "reminder deleted"})
}
